QuickCheck error

Running the parallel tests using `-quick-seed=36575` results in a consistency error after 1m23s. Failure happens consistently on Linux and Darwin.

See also: https://drone.io/github.com/boltdb/bolt/78

``` sh
$ go test -v -race -quick.seed=36575 -quick.count=20 -test.run=Parallel
seed: 36575
quick settings: count=20, items=1000, ksize=1024, vsize=1024
=== RUN TestParallelTxs
....................
--- FAIL: TestParallelTxs (125.70 seconds)
        Location:       functional_test.go:66
    Error:      Not equal: []byte{0x21, 0x5d, 0x41, 0xd2, 0x34, 0x25, 0x90, 0x89, 0x20, 0xb0, 0x71, 0x59, 0xe5, 0x65, 0x3a, 0x7d, 0x44, 0x56, 0x8d, 0xa8, 0xd8, 0x1f, 0x69, 0x4f, 0xe4, 0x35, 0x38, 0xf1, 0x9c, 0x2, 0xcc, 0x9b, 0x6b, 0xf, 0x72, 0xa3, 0x13, 0xd8, 0x88, 0xa1, 0x1b, 0x20, 0x17, 0xb5, 0x30, 0x82, 0x10, 0x35, 0x67, 0xe3, 0x1d, 0x1a, 0x2c, 0x10, 0xef, 0x3a, 0x9a, 0xda, 0x92, 0x2a, 0x70, 0xe7, 0xad, 0x6, 0x7b, 0x15, 0xec, 0x1d, 0x94, 0xb1, 0x5c, 0x9d, 0x79, 0x64, 0x62, 0x81, 0x2d, 0xf0, 0x98, 0x8e, 0x77, 0x74, 0x4d, 0xec, 0xd1, 0xdd, 0xeb, 0xc1, 0x42, 0x75, 0xc8, 0xa7, 0xfd, 0x16, 0xab, 0x70, 0xda, 0xf6, 0x7f, 0x2e, 0xfc, 0x28, 0x5a, 0x85, 0xd9, 0x94, 0xe, 0x4, 0x17, 0x3b, 0xeb, 0xa4, 0x2b, 0x64, 0x20, 0x6f, 0x86, 0xb, 0x3c, 0x20, 0x36, 0xaa, 0x16, 0xc7, 0x5c, 0xdc, 0x6a, 0xa1, 0xcc, 0x98, 0xaf, 0xc2, 0xb5, 0x1d, 0xef, 0x3b, 0x50, 0xf2, 0xb9, 0x43, 0x2e, 0x9f, 0xf9, 0x77, 0xb0, 0xcf, 0xd0, 0x70, 0x31, 0x80, 0xb7, 0xa3, 0x2e, 0x24, 0x65, 0x88, 0xa3, 0xa1, 0xd7, 0xbb, 0xb5, 0x44, 0x7c, 0x4e, 0xb0, 0x33, 0x4c, 0x4e, 0x8a, 0x2e, 0xa0, 0xf0, 0xc5, 0xce, 0xc6, 0x3c, 0x20, 0x6b, 0xab, 0xdc, 0xa3, 0x35, 0x16, 0x51, 0x25, 0xa, 0x6c, 0x31, 0x92, 0x22, 0x4a, 0xd9, 0xf0, 0x5c, 0x85, 0x8, 0xa7, 0xb2, 0x27, 0x10, 0x60, 0x79, 0x38, 0xe6, 0x93, 0x53, 0xf2, 0xd2, 0x48, 0x57, 0x59, 0x95, 0xe4, 0x64, 0x97, 0x8b, 0x67, 0x85, 0x6c, 0x18, 0xa7, 0x2f, 0xb2, 0x97, 0xab, 0xb6, 0xb7, 0x81, 0xfd, 0xba, 0x9e, 0xf9, 0x82, 0xda, 0x4a, 0x71, 0xea, 0x85, 0xbc, 0x9, 0x9f, 0x55, 0xee, 0x7b, 0x7b, 0x2b, 0x5a, 0x40, 0x8a, 0x93, 0xa5, 0x30, 0x21, 0xea, 0xa2, 0xd8, 0x25, 0xb2, 0xe1, 0x3e, 0xa6, 0x82, 0x4e, 0x9f, 0x94, 0xd4, 0x66, 0x92, 0x75, 0x62, 0xfc, 0xe1, 0x86, 0x7f, 0x7e, 0x43, 0x1e, 0xc1, 0xbf, 0x3f, 0xf5, 0x55, 0x9, 0x4, 0x98, 0xa3, 0x16, 0x17, 0x29, 0x35, 0x8b, 0xb0, 0xbd, 0x41, 0x9a, 0xda, 0x1d, 0x25, 0x33, 0xeb, 0xef, 0xd, 0x4a, 0x13, 0xb9, 0x90, 0xb0, 0x70, 0x30, 0xe7, 0x2c, 0x9a, 0x78, 0x6c, 0x8b, 0xd3, 0xa0, 0xf4, 0x80, 0xc8, 0x9, 0xb5, 0xbc, 0x8f, 0x71, 0x4, 0x2e, 0x3b, 0x7e, 0x1c, 0xe7, 0xee, 0x33, 0xce, 0x49, 0x85, 0x2b, 0xf1, 0x29, 0x7, 0x46, 0x44, 0x8e, 0xd5, 0xaa, 0xf9, 0x76, 0xd2, 0x72, 0xb5, 0x48, 0x52, 0x98, 0x21, 0x2, 0x9d, 0xd1, 0xc0, 0xb4, 0x60, 0x68, 0x93, 0xba, 0xbc, 0x5c, 0x5, 0x3d, 0xb0, 0x2, 0x3a, 0x77, 0x18, 0x79, 0x9a, 0xb4, 0x1d, 0x98, 0x97, 0x6f, 0x26, 0x31, 0xb2, 0xf9, 0x84, 0xee, 0x8f, 0xf9, 0xf4, 0x94, 0xae, 0xd8, 0x7, 0x4d, 0xc1, 0xa, 0x5d, 0xd5, 0x86, 0x1a, 0x91, 0x7a, 0x32, 0x1a, 0x2f, 0x3f, 0xb1, 0x51, 0x23, 0x7e, 0xb4, 0x15, 0x1e, 0xef, 0xe2, 0xc0, 0x2f, 0x85, 0x4a, 0x4e, 0x9a, 0xdf, 0xb9, 0x70, 0x96, 0xf, 0xc4, 0xf7, 0x81, 0x8, 0xa4, 0x28, 0x9, 0x56, 0xd8, 0xc0, 0x34, 0xf8, 0xde, 0xbd, 0x9c, 0x16, 0x3b, 0xeb, 0x16, 0xf6, 0xc6, 0xeb, 0x69, 0x1f, 0x99, 0x51, 0xa7, 0x40, 0xa4, 0x7, 0x43, 0x77, 0xea, 0xe, 0x72, 0x4e, 0x37, 0xe5, 0x89, 0x72, 0xd5, 0x3, 0x30, 0x2c, 0x31, 0x76, 0xb0, 0xcb, 0x96, 0x60, 0xd9, 0x24, 0xa6, 0x16, 0x81, 0x3, 0xd8, 0xc5, 0xc5, 0xb6, 0xc2, 0xd1, 0xbe, 0xd2, 0x32, 0x26, 0x7b, 0xdc, 0x73, 0xe3, 0xe1, 0x5f, 0x23, 0x4, 0x8b, 0xb0, 0x55, 0x3e, 0x26, 0xdb, 0x1e, 0x84, 0xcb, 0x70, 0x43, 0xf2, 0x10, 0x45, 0x8d, 0x4f, 0x8e, 0xfe, 0xf0, 0x6b, 0x46, 0x8f, 0x8f, 0x32, 0xf0, 0xfb, 0x3b, 0x2b, 0x47, 0xf6, 0x84, 0xe9, 0xab, 0xf3, 0x84, 0xfa, 0x12, 0xae, 0xe0, 0x5c, 0xaa, 0x10, 0xaf, 0x72, 0x8f, 0x85, 0xb6, 0xf3, 0xb1, 0x51, 0x68, 0xdf, 0xb3, 0xd4, 0xd0, 0x7a, 0x25, 0x3b, 0x90, 0xb9, 0xce, 0x91, 0x2a, 0x17, 0x4e, 0xd8, 0x6f, 0x4e, 0xaa, 0x85, 0xab, 0xc6, 0x80, 0x11, 0x70, 0x28, 0x12, 0x66, 0x7, 0xa7, 0xa2, 0x56, 0x27, 0x27, 0x26, 0x65, 0xed, 0x9f, 0xed, 0x1d, 0x27, 0xca, 0xb3, 0xcd, 0xaf, 0x13, 0x2c, 0x86, 0xb7, 0xd5, 0x56, 0x50, 0xc3, 0x67, 0x34, 0xec, 0x8f, 0xc3, 0x6b, 0xe3, 0x6d, 0xc5, 0x5e, 0x2c, 0xd6, 0xce, 0x2e, 0x87, 0x9b, 0xd3, 0x5d, 0xb8, 0x2c, 0xb2, 0xd, 0xa8, 0x27, 0xca, 0x5b, 0x3d, 0xba, 0x11, 0x1a, 0x15, 0xa1, 0x5, 0x58, 0xbf, 0x29, 0xa0, 0xa8, 0xbf, 0x74, 0x5d, 0xb, 0xf0, 0x3, 0xbb, 0x4f, 0x3e, 0xaf, 0xcc, 0x20, 0x28, 0xc1, 0xe7, 0x61, 0xe8, 0x8c, 0x2a, 0x27, 0x5, 0x84, 0xe8, 0x73, 0x61, 0x5f, 0x6b, 0x7b, 0xc4, 0x3, 0x8d, 0x95, 0x27, 0x67, 0x50, 0xaf, 0xe0, 0x9b, 0xcf, 0x4b, 0x88, 0xe0, 0x36, 0x42, 0xf, 0x8, 0x71, 0xb8, 0xa3, 0x53, 0x37, 0x21, 0x79, 0x4c, 0xfb, 0x33, 0x8c, 0xd8, 0x6d, 0x75, 0x98, 0xb4, 0x94, 0x3, 0x7f, 0xd5, 0x8, 0x3d, 0xdb, 0xd3, 0xe3, 0x98, 0x31, 0x24, 0xbf, 0x83, 0xd1, 0x62, 0x6c, 0x18, 0xce, 0xbe, 0xa, 0xd7, 0xb3, 0x7c, 0xc5, 0xbd, 0x6d, 0x8e, 0x33, 0xe9, 0xce, 0xb9, 0x59, 0x33, 0xa1, 0xa2, 0x19, 0xb9, 0x4e, 0x9b, 0x6d, 0x47, 0xa4, 0x5, 0x59, 0xe, 0x7a, 0x9e, 0x9d, 0xc7, 0xd0, 0xd4, 0x49, 0xf6, 0xc5, 0xee, 0x67, 0x80, 0xe3, 0x39, 0xba, 0xca, 0x9d, 0xb5, 0xaf, 0x19, 0xa5, 0xdb, 0x72, 0x35, 0x3e, 0xfc, 0xa3, 0xc8, 0x54, 0x7d, 0x8e, 0xca, 0x9b, 0x37, 0x0, 0x92, 0xfa, 0x71, 0xe9, 0xf3, 0xc3, 0x95, 0x43, 0x10, 0x51, 0x7b, 0x40, 0xc4, 0x13, 0xc5, 0x95, 0x1c, 0x23, 0xfb, 0xc, 0x91, 0xde, 0x1c, 0x94, 0xdb, 0xb, 0xe6, 0x48, 0x70, 0x7e, 0x4, 0x82, 0xbb, 0x67, 0x5e, 0xcb, 0x59, 0x95, 0x69, 0x73, 0xc6, 0xb9, 0xc1, 0xe9, 0x22, 0xd4, 0x6e, 0x29, 0x65, 0x6, 0xc9, 0x2c, 0x94, 0x9, 0x77, 0x6a, 0x51, 0x6, 0x50, 0x9c, 0x32, 0x7, 0x4e, 0xf9, 0x47, 0xe6, 0xd5, 0x82, 0x24, 0xcd, 0x30, 0xe7, 0x32, 0xb9, 0xb, 0x1e, 0xc1, 0x58, 0xc0, 0x80, 0x70, 0x67, 0xb0, 0x6d, 0x39, 0xaf, 0x55, 0x57, 0xe9, 0xaa, 0xb7, 0xd6, 0x5c, 0x16, 0x69, 0xf0, 0x4a, 0x6a, 0xe2, 0x30, 0xe4, 0x9f, 0xeb, 0x16, 0xc4, 0x84, 0x91, 0xa0, 0x6c, 0x80, 0x4e, 0xc3, 0x5c, 0x42, 0x35, 0x57, 0x8, 0x1c, 0xd7, 0x96, 0x2f, 0x40, 0x49, 0xd, 0xf4, 0xe5, 0x8a, 0xbc, 0x5b, 0xb9, 0x71, 0xe4, 0xd, 0xf3, 0x1e, 0x1d, 0xb4, 0xbc, 0xc9, 0x84, 0x6d, 0x94, 0xa9, 0xa, 0x70, 0x96, 0xa5, 0x1c, 0x9b, 0x2c, 0xb3, 0xbf, 0xe6, 0x90, 0x4a, 0xcd, 0x1f, 0xca, 0x8e, 0x89, 0x39, 0x8, 0x50, 0x6f, 0x96, 0xf4, 0xef, 0x72, 0x7a, 0x45, 0x7f, 0x33, 0x28, 0x77, 0x57, 0x5a, 0x8a, 0x5a, 0x26, 0x71, 0x88, 0xe2, 0xae, 0xac, 0x65, 0xc6, 0x47, 0x16, 0x49, 0xfa, 0x56, 0x65, 0x96, 0xbd, 0x77, 0xee, 0x81, 0xd8, 0x10, 0xf1, 0x7d, 0x30, 0xa0, 0xca, 0xf1, 0x36, 0xcb, 0xe7, 0xdb, 0x27, 0x49, 0xb4, 0x1e, 0x6, 0xc, 0x3, 0x92, 0x7f, 0xf7, 0x6c, 0x9, 0xf0, 0xe9} != []byte{0x3, 0xc8, 0x89, 0xa4, 0x90, 0xe, 0x2f, 0xd7, 0x88, 0xda, 0x86, 0xa4, 0x41, 0x5b, 0x6c, 0xa7, 0x5f, 0xfd, 0xdd, 0x60, 0x44, 0x11, 0x24, 0x6f, 0x5a, 0x7c, 0x60, 0x9b, 0xb7, 0x8e, 0x1b, 0x9c, 0xe1, 0x48, 0x3e, 0x7a, 0xd7, 0xeb, 0xa1, 0x8f, 0x31, 0xa8, 0xd3, 0x0, 0x78, 0x84, 0x22, 0x4e, 0x8d, 0xf7, 0x98, 0xf2, 0x25, 0x3, 0xa4, 0x1b, 0x3d, 0xaf, 0xbd, 0x92, 0xef, 0xd5, 0x2f, 0x9f, 0x99, 0x81, 0xb4, 0x31, 0x6a, 0x41, 0x72, 0xf4, 0x26, 0xd4, 0xa9, 0x5e, 0x18, 0x54, 0x17, 0x1b, 0x4a, 0xde, 0xba, 0x75, 0x54, 0x1b, 0xbf, 0x4f, 0x5b, 0xcc, 0xb6, 0xa6, 0x77, 0xdf, 0xf7, 0x90, 0xda, 0x24, 0xb8, 0xeb, 0x17, 0xc0, 0xa2, 0x5e, 0xfa, 0xb0, 0xf2, 0x60, 0xb9, 0x82, 0xe, 0xc3, 0xe3, 0xf2, 0xa4, 0x5a, 0x6d, 0x2e, 0xe8, 0x7b, 0x4, 0xb3, 0x6e, 0xf9, 0x37, 0xbf, 0x2e, 0xf3, 0xc7, 0x5e, 0x70, 0x50, 0x91, 0xe8, 0x2a, 0xc0, 0x15, 0x7d, 0xb7, 0x67, 0x8a, 0xfb, 0xa7, 0xa0, 0x6a, 0xae, 0xa4, 0x76, 0xbb, 0xb9, 0xf6, 0xdd, 0x1f, 0xc1, 0x51, 0xb0, 0xd8, 0x64, 0xa4, 0xd9, 0x7a, 0x8, 0xdb, 0x3d, 0xef, 0x5e, 0xb3, 0x30, 0xbb, 0xb3, 0x24, 0x50, 0x64, 0x2c, 0xd7, 0x74, 0xd8, 0x72, 0xb0, 0xb4, 0xc1, 0x85, 0xbc, 0x4a, 0xd2, 0xf1, 0xcf, 0xa3, 0xda, 0x76, 0xe1, 0xbe, 0xb3, 0x15, 0x28, 0x20, 0x85, 0xb6, 0xbf, 0x5d, 0x98, 0x26, 0x1a, 0x11, 0xb, 0x21, 0x95, 0xc7, 0xfb, 0x5d, 0x62, 0x66, 0x87, 0x5b, 0x91, 0x2d, 0xf6, 0xce, 0x0, 0xb7, 0x4a, 0xd3, 0x2f, 0x41, 0x3b, 0xcb, 0xb8, 0x6c, 0x95, 0xd1, 0x35, 0xbd, 0x30, 0xf0, 0x82, 0x37, 0x66, 0x94, 0x1, 0x22, 0xb7, 0xfd, 0xb2, 0xe3, 0x13, 0x33, 0x3a, 0x65, 0xca, 0xfe, 0xb8, 0x28, 0xdb, 0xfc, 0x28, 0x73, 0x67, 0xf1, 0x81, 0x55, 0xec, 0xe6, 0xc5, 0x9f, 0xa8, 0xf3, 0xcf, 0xd7, 0x14, 0xb9, 0xe0, 0x8, 0xdc, 0x64, 0x3, 0xac, 0xac, 0x6b, 0x30, 0x70, 0xe7, 0x64, 0x9b, 0x7d, 0x9e, 0xef, 0x6e, 0xf8, 0x9, 0x79, 0xf4, 0xea, 0x1c, 0x2f, 0xe3, 0xb7, 0xe9, 0xa3, 0x27, 0x61, 0xdd, 0xd3, 0x9, 0xfa, 0xa7, 0x7f, 0x44, 0x76, 0x52, 0xa9, 0x5, 0xb8, 0xd1, 0x32, 0xbd, 0xb1, 0xf8, 0x26, 0x81, 0xa9, 0x7b, 0x78, 0xf3, 0xf6, 0xe9, 0x42, 0x32, 0xe2, 0xb3, 0xc3, 0x4a, 0xcf, 0x3b, 0xc9, 0x37, 0xcd, 0xf3, 0xbe, 0xce, 0xbd, 0x64, 0xe9, 0xaa, 0xc1, 0xbd, 0x21, 0xca, 0x21, 0x9a, 0x7a, 0x88, 0x53, 0x58, 0xe8, 0xd0, 0x9, 0xd3, 0x3, 0xab, 0x86, 0x19, 0x21, 0x3a, 0xfd, 0x78, 0xb2, 0x88, 0xcf, 0x22, 0x13, 0x16, 0x66, 0xe6, 0x32, 0x79, 0xa4, 0xf, 0xed, 0x80, 0xd4, 0xf6, 0x51, 0x59, 0xab, 0xde, 0x9, 0x6, 0x1e, 0x85, 0x28, 0x13, 0x85, 0xe, 0x88, 0x8, 0x90, 0xaf, 0xa7, 0x6b, 0x9d, 0xd3, 0xa9, 0x3d, 0xa2, 0xf2, 0xf3, 0xf3, 0xda, 0xad, 0x5e, 0xd5, 0x15, 0xbe, 0x7e, 0x75, 0xc8, 0x1a, 0xee, 0xca, 0x4, 0xd5, 0xec, 0x0, 0x54, 0x6c, 0xa8, 0xbe, 0xa1, 0xbb, 0x41, 0x20, 0x85, 0x9c, 0x90, 0xbb, 0x55, 0x62, 0x6e, 0x13, 0xa1, 0x9a, 0x91, 0x6e, 0x22, 0x1e, 0x48, 0x58, 0x6a, 0x84, 0x88, 0xd, 0x63, 0x3d, 0x71, 0x73, 0xb5, 0x82, 0x18, 0xda, 0xd0, 0x4a, 0x45, 0x6e, 0x31, 0x6f, 0xa7, 0xbf, 0x9f, 0x18, 0x50, 0xda, 0x19, 0x6e, 0x5c, 0x20, 0x54, 0x66, 0xdf, 0x86, 0x55, 0x21, 0xcb, 0x89, 0x48, 0x74, 0x82, 0x6d, 0x8d, 0xf4, 0xef, 0x76, 0x96, 0x4d, 0x41, 0x8a, 0xdb, 0xd, 0x37, 0xe3, 0x75, 0x5a, 0xf4, 0xc3, 0x68, 0xd8, 0x61, 0x7c, 0xc1, 0xce, 0x67, 0xa6, 0x34, 0xf7, 0x57, 0x41, 0x77, 0x52, 0x74, 0x1a, 0xf1, 0x4, 0xed, 0x5f, 0x9, 0x49, 0x72, 0x33, 0x6c, 0x8a, 0x19, 0xb3, 0xa0, 0x5a, 0x9e, 0x7a, 0xfb, 0x61, 0x2, 0xe9, 0xeb, 0x3a, 0x52, 0xda, 0xc2, 0x4a, 0xee, 0xde, 0x10, 0xe4, 0x66, 0xbb, 0x50, 0x8a, 0xa4, 0x78, 0x6b, 0xd1, 0xb4, 0x1a, 0x75, 0x53, 0x7e, 0xda, 0x65, 0x38, 0xc6, 0x37, 0xb2, 0xfd, 0xaf, 0xb2, 0xae, 0x5c, 0xf6, 0x32, 0xd3, 0x1e, 0xb7, 0xe1, 0x7, 0x5f, 0x57, 0x9b, 0xb3, 0x4, 0xf3, 0x5f, 0xad, 0x1, 0xb7, 0x6a, 0xe, 0x4f, 0x64, 0x1a, 0x4c, 0xfd, 0x7c, 0xa, 0x4a, 0xb2, 0xb9, 0xa2, 0xb8, 0xea, 0x18}
...
```

--- a/functional_test.go
+++ b/functional_test.go
@@ -24,6 +24,10 @@ func TestParallelTxs(t *testing.T) {
 		numReaders = (numReaders % 10) + 1
 		batchSize = (batchSize % 50) + 1
 
+		// warn("")
+		// warn("================================================================")
+		// warn("numReaders:", numReaders, "batchSize", batchSize)
+
 		// Maintain the current dataset.
 		var current testdata
 
@@ -50,7 +54,10 @@ func TestParallelTxs(t *testing.T) {
 
 					go func() {
 						mutex.RLock()
-						local := current
+						local := make(map[string][]byte)
+						for _, item := range current {
+							local[string(item.Key)] = item.Value
+						}
 						tx, err := db.Begin(false)
 						mutex.RUnlock()
 						if err == ErrDatabaseNotOpen {
@@ -61,9 +68,9 @@ func TestParallelTxs(t *testing.T) {
 						}
 
 						// Verify all data is in for local data list.
-						for _, item := range local {
-							value := tx.Bucket("widgets").Get(item.Key)
-							if !assert.NoError(t, err) || !assert.Equal(t, value, item.Value) {
+						for k, v := range local {
+							value := tx.Bucket("widgets").Get([]byte(k))
+							if !assert.NoError(t, err) || !assert.Equal(t, value, v, fmt.Sprintf("reader (%p)", tx)) {
 								tx.Rollback()
 								wg.Done()
 								t.FailNow()
@@ -94,9 +101,12 @@ func TestParallelTxs(t *testing.T) {
 					t.FailNow()
 				}
 
+				// warnf("[writer] BEGIN (%d)", currentBatchSize)
+
 				// Insert whole batch.
 				b := tx.Bucket("widgets")
 				for _, item := range batchItems {
+					// warnf("[writer] PUT %x: %x", trunc(item.Key, 3), trunc(item.Value, 3))
 					err := b.Put(item.Key, item.Value)
 					if !assert.NoError(t, err) {
 						t.FailNow()
@@ -105,6 +115,7 @@ func TestParallelTxs(t *testing.T) {
 
 				// Commit and update the current list.
 				mutex.Lock()
+				// warnf("[writer] COMMIT\n\n")
 				err = tx.Commit()
 				current = append(current, batchItems...)
 				mutex.Unlock()

